<a href="https://colab.research.google.com/github/HugoCortez01/ImageModel/blob/main/cc_apc_2026.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **`Algoritmos e Pensamento Computacional`**
---
---

## **Revisão:**
---
---
+ ## Estrutura lógica sequencial (bloco de comandos).
    + ### Variáveis e Comandos de Entrada e Saída de dados.
    + ### Comando de atribuição, cálculos e funções matemáticas.
+ ## Estrutura lógica condicional (simples e composta).
    + ### Operadores relacionais e operadores lógicos.

## **Estrutura lógica sequencial (bloco de comandos)**:
---
+ ### Variáveis:
    + #### Regra para nomes de variáveis: `letra ou _, seguido ou não de letras, dígitos ou _`.

In [ ]:
a = 10
b = 20
soma = a + b
print('Soma =', soma)

Soma = 30


### **Operadores matemáticos:**
---

+ ### Adição: `+`
+ ### Subtração: `-`
+ ### Multiplicação: `*`
+ ### Divisão: `/`
+ ### Resto da divisão: `%`
+ ### Quociente da divisão: `//`
+ ### Potência: `**`

### **Prioridade**: `**`, `* / % //` e, por último, `+ -`.
Operações de mesma prioridade são executadas da esquerda para a direita, com exceção da potência que é executada da direita para a esquerda.

In [ ]:
print(5 + 7)
print(5 - 9)
print(3 * 6)
print(5 / 2)
print(11 % 3)
print(11 // 3)
print(2 ** 10)

12
-4
18
2.5
2
3
1024


In [ ]:
# área de um retângulo

lado1 = float(input('Lado 1 (m): '))  # entrada
lado2 = float(input('\nLado 2 (m): '))  # entrada
area = lado1 * lado2
print(f'\nÁrea do retângulo = {lado1:.2f} x {lado2:.2f} = {area:.2f} m²')  # saída

Lado 1 (m): 1.56

Lado 2 (m): 2.695

Área do retângulo = 1.56 x 2.69 = 4.20 m²


### **Funções matemáticas:**
---
+ ### Módulo (biblioteca) `math` (import).

In [ ]:
# Exemplo de cálculo da hipotenusa de um triângulo retângulo

from math import sqrt # use 'import math' para importar o módulo inteiro

ca = float(input('Cateto A: '))
cb = float(input('\nCateto B: '))
hi = sqrt(ca ** 2 + cb ** 2)  # Ou ainda: (ca ** 2 + cb ** 2) ** 0.5
print(f'\nHipotenusa = {hi:.3f}')

Cateto A: 3

Cateto B: 4

Hipotenusa = 5.000


In [ ]:
# Exemplo de cálculo do seno de um ângulo em graus

from math import sin, radians

ang = float(input('Ângulo (graus): '))
seno = sin(radians(ang))  # ou converter usando a fórmula: ang * 3.14 / 180
print(f'\nSeno de {ang:.1f}° = {seno:.3f}')

Ângulo (graus): 30

Seno de 30.0° = 0.500


## **Estrutura lógica condicional (simples e composta):**
---
### `Algoritmo`: comando simples e composto.
    se ? então
    início
        ...comandos...   (verdadeiro)
    fim
#  
    se ? então
    início
        ...comandos...   (verdadeiro)
    fim
    senão
    início
        ...comandos...   (falso)
    fim


### `?` : condição (V ou F).

 ## **Linguagem `Python`:** o deslocamento para direita (Tab) inicia um bloco de comandos e o retorno para a coluna à esquerda, finaliza o bloco.
 ---

    if ?:       # Comando condicional SIMPLES
        ...
        ...
#  
    if ?:       # Comando condicional COMPOSTO
        ...
        ...
    else:
        ...
        ...


### `?` : condição (V ou F).

## **`Condições`:**
---
+ ### Resultam sempre em Verdadeiro ou Falso (`True` ou `False`).
+ ### Operadores relacionais:
 + ## `<` `<=` `>` `>=` `==` `!=`
+ ### Operadores lógicos:
 + ## `and` (e) `or` (ou) `not` (negação)

### **Prioridade (ou precedência) de Operadores:**

[https://www.inf.pucrs.br/flash/algfis/aulas/selecao/build/selecao/PrecedenceofOperators.html](https://www.inf.pucrs.br/flash/algfis/aulas/selecao/build/selecao/PrecedenceofOperators.html)

## **Exemplos:**
---
---

### `1.` Programa para mostrar o maior valor entre 3 números inteiros:

In [ ]:
a = int(input('Valor de A: '))
b = int(input('\nValor de B: '))
c = int(input('\nValor de C: '))

if a > b and a > c:
    ma = a
elif b > c:
    ma = b
else:
    ma = c

print(f'\nMaior = {ma}')

Valor de A: 5

Valor de B: 6

Valor de C: 4

Maior = 6


### `Outra forma`: algoritmo mais eficiente.

In [ ]:
a = int(input('Valor de A: '))
b = int(input('\nValor de B: '))
c = int(input('\nValor de C: '))

ma = a
if b > ma:
    ma = b
if c > ma:
    ma = c

print(f'\nMaior = {ma}')

Valor de A: 4

Valor de B: 5

Valor de C: 1

Maior = 5


### `2.` Criar um programa para ler 5 números reais (float) e, após a leitura, calcular e mostrar a qtde de valores negativos entre os 5 números.

+ ### Usar um `contador`.

In [ ]:
n1 = float(input('Número 1: '))
n2 = float(input('Número 2: '))
n3 = float(input('Número 3: '))
n4 = float(input('Número 4: '))
n5 = float(input('Número 5: '))

qn = 0  # qtde de negativos (contador)
if n1 < 0:
    qn += 1  # qn = qn + 1 (incrementar)
if n2 < 0:
    qn += 1
if n3 < 0:
    qn += 1
if n4 < 0:
    qn += 1
if n5 < 0:
    qn += 1

print(f'\nQtde de negativos = {qn}')

Número 1: 4
Número 2: -8
Número 3: -9
Número 4: 6
Número 5: -8

Qtde de negativos = 3


## **Estrutura Lógica de Repetição:**
---
+ ## Comandos de repetição:
    + ## `while`
    + ## `for`

### Algoritmo (pseudocódigo) dos comandos enquanto (`while`) e para (`for`):

    enquanto ? faça
    início
        ...comandos...
    fim
#  
    para n de 1 até 10 faça
    início
        ...comandos...
    fim

## **Exemplos:**
---

### `1.` Criando uma repetição infinita.

In [ ]:
while True:
    print('$', end='')

### `2.` Criando uma repetição finita (contador).

In [ ]:
cont = 0
while cont < 20:
    print(cont, end=' ')
    cont += 1  # cont = cont + 1

0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 

### `Outra forma`: usando o comando `break`.

In [ ]:
cont = 0
while True:
    print(cont, end=' ')
    cont += 1
    if cont > 19:
        break  # termina a execução do while e desvia para o comando após o while

0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 

### `3.` Usando o `for` para criar uma repetição finita (mesmo resultado do anterior).

In [ ]:
for cont in range(20):
    print(cont, end=' ')

0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 

### `4.` Criar um programa para somar os números inteiros de um intervalo entre **`i`** (início) e **`f`** (fim) (inclusive i e f), com **i** e **f** fornecidos pelo usuário.
+ ### Contador e Acumulador

In [ ]:
# usando o while

i = int(input('Início: '))
f = int(input('\nFim: '))

c = i  # c é o contador e i é o início do intervalo
s = 0  # s é o acumulador (0 é o elemento neutro da soma)
while c <= f:  # f é o fim do intervalo
    s += c  # s = s + c (s é o acumulador)
    c += 1  # incrementa o contador

print(f'\nA soma entre {i} e {f} é igual a {s}')

Início: 1

Fim: 1000

A soma entre 1 e 1000 é igual a 500500


### `Outra forma`: usando o for.

In [ ]:
# usando o for

i = int(input('Início: '))
f = int(input('\nFim: '))

s = 0  # soma (acumulador)
for c in range(i, f + 1):
    s += c  # s = s + c

print(f'\nA soma entre {i} e {f} é igual a {s}')

Início: 1

Fim: 1000

A soma entre 1 e 1000 é igual a 500500


### `5.` Criar uma programa para ler uma certa qtde de valores reais (`float`) e, após as leituras, mostrar a média aritmética dos números digitados. A qtde de valores é escolhida pelo usuário no início do programa.
### Ex.:
    Qtde de valores: 5
    Valor 1: 6
    Valor 2: 2
    Valor 3: 7
    Valor 4: 5
    Valor 5: 1
    Média aritmética = 4.20

In [ ]:
qv = int(input('Qtde de valores: '))
print()  # salta uma linha
s = 0  # armazena a soma dos valores (acumulador)
for i in range(qv):  # repete qv vezes
    val = float(input(f'Valor {i + 1}: '))  # leitura dos valores
    s += val  # soma dos valores

ma = s / qv  # cálculo da média
print(f'\nMédia aritmética = {ma:.2f}')

Qtde de valores: 5

Valor 1: 6
Valor 2: 2
Valor 3: 7
Valor 4: 5
Valor 5: 1

Média aritmética = 4.20


# **Arrays** (Vetores): uma dimensão.
---
+ ### Listas numéricas.

## Criação, alteração e seleção de valores em listas.
---

In [ ]:
lista = []  # criar uma lista vazia
print(lista, type(lista))

[] <class 'list'>


In [ ]:
lista = [1, 2, 3, 4, 5]
print(lista)

[1, 2, 3, 4, 5]


In [ ]:
lista = [0] * 30  # criar uma lista com 30 zeros
print(lista)

[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]


In [ ]:
lista = []
for v in range(30):
    lista.append(v + 1)  # append acrescenta um valor no final da lista
print(lista)

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30]


In [ ]:
lista = [v + 1 for v in range(30)]  # o for fornece os valores da lista
print(lista)

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30]


In [ ]:
# alterações em posições (índices) de arrays

lista = [0] * 10
print(f'Antes: {lista}')
lista[0] = 111  # altera o primeiro
lista[-1] = 999  # altera o último
print(f'\nDepois: {lista}')

Antes: [0, 0, 0, 0, 0, 0, 0, 0, 0, 0]

Depois: [111, 0, 0, 0, 0, 0, 0, 0, 0, 999]


In [ ]:
# recortes em listas numéricas (arrays)

lista = [v for v in range(10)]
print(lista)  # lista inteira
print(lista[5:])  # do 6º até o fim
print(lista[:5])  # do 1º ao 5º
print(lista[2:5])  # do 3º ao 5º
print(lista[-3:])  # os 3 últimos
print(lista[::2])  # valores pares (passo 2)
print(lista[1::2])  # valores ímpares (passo 2)

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
[5, 6, 7, 8, 9]
[0, 1, 2, 3, 4]
[2, 3, 4]
[7, 8, 9]
[0, 2, 4, 6, 8]
[1, 3, 5, 7, 9]


## **Resumo das operações com listas** (métodos e funções):
---

In [ ]:
lis = []  # criar uma lista vazia
lis = [1, 2, 3, 4, 5]  # criar uma lista com valores pré-inseridos
lis = [0] * 10  # criar uma lista com 10 zeros
lis = [x for x in range(10)]  # criar a lista: [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
copia = lis.copy()  # copiar a lista lis para copia
copia = lis[:]  # outra forma de copiar a lista lis para copia
print(copia)
print(copia[1:4])  # mostra o recorte da lista entre o 2º e o 4º valor
lis.append(10)  # inserir no final da lista o valor 10
print(lis)
val = lis.pop()  # retirar da lista o valor final (val recebe o último da lista)
print(val, lis)
val = lis.pop(0)  # retirar o primeiro da lista (o índice deve ser passado)
print(val, lis)
lis.insert(4, 99)  # inserir 99 na posição (índice) 4 (5º posição)
print(lis)
lis = [6, 2, 9, 0, 2, 6, 7, 3]
print(f'Lista desordenada = {lis}')
tam = len(lis)  # calcular o tamanho da lista (nº de elementos)
print(tam, 'elementos')
novalis = sorted(lis)  # criar uma nova lista ordenada (crescente) da lista passada com parâmetro (lis)
print(f'Lista desordenada = {lis} e lista em ordem crescente = {novalis}')
novalis = sorted(lis, reverse=True)  # ordem decrescente
print(f'Lista desordenada = {lis} e lista em ordem decrescente = {novalis}')
lis.sort()  # ordenar a lista (a lista será alterada)
print(f'Lista ordenada = {lis}')
me, ma, soma = min(lis), max(lis), sum(lis)  # calcular menor, maior e a soma dos valores de uma lista
print('Menor =', me, 'Maior =', ma, 'Soma =', soma)
cont = lis.count(2)  # calcular a qtde de elementos iguais ao 2
print('Qtde iguais ao 2 =', cont)
if 3 in lis:  # verifica se o valor 3 está contido na lista lis
    print('3 está contido na lista')
if 5 not in lis:  # verifica se o 5 NÃO está contido na lista lis
    print('SIM')
else:
    print('NÃO')
lis1 = [1, 2, 3]
lis2 = [4, 5, 6, 7, 8]
lis3 = lis1 + lis2  # soma de listas resulta em outra lista com os valores reunidos
print(f'Lista (listas somadas): {lis3}')
lis = [3, 4, 2, 1, 5, 3, 4, 2, 4]
print(lis)
lis = list(set(lis))  # eliminar repetições em lista (set converte em conjunto)
print(lis)
lis = []
if not lis:  # lista vazia é False
    print('lista vazia')

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
[1, 2, 3]
[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
10 [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
0 [1, 2, 3, 4, 5, 6, 7, 8, 9]
[1, 2, 3, 4, 99, 5, 6, 7, 8, 9]
Lista desordenada = [6, 2, 9, 0, 2, 6, 7, 3]
8 elementos
Lista desordenada = [6, 2, 9, 0, 2, 6, 7, 3] e lista em ordem crescente = [0, 2, 2, 3, 6, 6, 7, 9]
Lista desordenada = [6, 2, 9, 0, 2, 6, 7, 3] e lista em ordem decrescente = [9, 7, 6, 6, 3, 2, 2, 0]
Lista ordenada = [0, 2, 2, 3, 6, 6, 7, 9]
Menor = 0 Maior = 9 Soma = 35
Qtde iguais ao 2 = 2
3 está contido na lista
SIM
Lista (listas somadas): [1, 2, 3, 4, 5, 6, 7, 8]
[3, 4, 2, 1, 5, 3, 4, 2, 4]
[1, 2, 3, 4, 5]
lista vazia


## **Exemplos:**
---

### `1.` Gerar uma lista com 10 números inteiros aleatórios entre 1 e 100 e, em seguida, separar os valores pares e ímpares em duas listas. Mostrar as 3 listas no final.

In [ ]:
from random import randint

lista = [randint(1, 100) for _ in range(10)]

par, impar = [], []
for v in lista:
    par.append(v) if v % 2 == 0 else impar.append(v)

# par = [v for v in lista if v % 2 == 0]
# impar = [v for v in lista if v % 2 != 0]

print(f'Lista: {lista}')
print(f'\nPares: {par}')
print(f'\nÍmpares: {impar}')

Lista: [2, 31, 85, 90, 43, 92, 19, 38, 80, 65]

Pares: [2, 90, 92, 38, 80]

Ímpares: [31, 85, 43, 19, 65]


### `2.` Gerar uma lista com 20 valores aleatórios entre 10 e 20 e, em seguida, ordenar a lista, calcular e mostrar (em ordem crescente) a qtde de ocorrências de cada valor dentro da lista (não repetir valores já mostrados).

In [ ]:
from random import randint

lis = [randint(10, 20) for _ in range(20)]
lis.sort()
print(f'Lista: {lis}\n')

print('Valores e quantidades:\n')
sel = []  # lista de valores já selecionados (mostrados)
for v in lis:
    if v not in sel:
        cont = 0
        for x in lis:
            if x == v:
                cont += 1
        print(f'{v}: {cont}x')
        sel.append(v)

print('\nValores e quantidades:\n')
sel = []  # lista de valores já selecionados (mostrados)
for v in lis:
    if v not in sel:
        print(f'{v}: {lis.count(v)}x')
        sel.append(v)

# Outra forma
print('\nValores e quantidades:\n')
for v in sorted(set(lis)):  # set (conjunto) elimina repetições em lis
    print(f'{v}: {lis.count(v)}x')

Lista: [10, 10, 11, 12, 13, 14, 15, 15, 16, 17, 17, 17, 18, 18, 18, 18, 19, 19, 19, 20]

Valores e quantidades:

10: 2x
11: 1x
12: 1x
13: 1x
14: 1x
15: 2x
16: 1x
17: 3x
18: 4x
19: 3x
20: 1x

Valores e quantidades:

10: 2x
11: 1x
12: 1x
13: 1x
14: 1x
15: 2x
16: 1x
17: 3x
18: 4x
19: 3x
20: 1x

Valores e quantidades:

10: 2x
11: 1x
12: 1x
13: 1x
14: 1x
15: 2x
16: 1x
17: 3x
18: 4x
19: 3x
20: 1x


### `3.` Criar um programa para gerar uma lista com `10 valores` inteiros **distintos** e aleatórios entre 1 e 100. Depois, gerar outra lista com `20 valores` aleatórios entre 1 e 100, também **distintos**. Após a criação das listas, ordená-las em ordem crescente e criar uma terceira lista com os valores presentes em ambas as listas (1ª e 2ª), ou seja, presente tanto na primeira como na segunda lista. Mostrar as 3 listas no final do programa.

In [ ]:
from random import sample

lis1 = sample(range(1, 101), 10)  # lista com 10 números distintos entre 1 e 100
lis1.sort()
lis2 = sample(range(1, 101), 20)  # lista com 20 números
lis2.sort()

lis3 = []
for v in lis1:
    if v in lis2:
        lis3.append(v)

print(f'Lista 1: {lis1}')
print(f'\nLista 2: {lis2}')
print(f'\nLista 3 = {lis3}')

Lista 1: [4, 9, 13, 35, 37, 39, 47, 69, 96, 97]

Lista 2: [1, 2, 4, 9, 13, 19, 29, 31, 32, 37, 38, 40, 48, 51, 54, 58, 67, 73, 83, 96]

Lista 3 = [4, 9, 13, 37, 96]


# **Arrays (Matrizes)**: duas ou mais dimensões.
---
+ ### Listas de listas numéricas.
+ ### Arrays NumPy.

## Criar e mostrar matrizes:

In [ ]:
# Criar uma matriz 3x4 (3 linhas e 4 colunas) com zeros

mat = [[0, 0, 0, 0],
       [0, 0, 0, 0],
       [0, 0, 0, 0]]
print(f'Matriz = {mat}')

Matriz = [[0, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0]]


In [ ]:
# Outra forma

mat = [[0] * 4 for i in range(3)]  # criar uma matriz 3x4 (i é a qtde de linhas)
print(f'Matriz = {mat}')

Matriz = [[0, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0]]


In [ ]:
# O mesmo com NumPy

import numpy as np

mat = np.zeros((3, 4))
print(mat)

[[0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]]


In [ ]:
# Mostrar uma matriz linha por linha

mat = [[9] * 3 for i in range(5)]

print('Matriz:')
for i in range(5):
    print(mat[i])  # i é a linha

print('\nMatriz:')
for lin in mat:  # mostra a matriz linha por linha
    print(lin)

print('\nMatriz:')
for i in range(5):  # i é linha
    for j in range(3):  # j é coluna
        print(mat[i][j], end=' ')
    print()

print('\nMatriz:')
for lin in mat:  # lin recebe uma linha
    for v in lin:  # v recebe valor por valor da linha (lin)
        print(v, end=' ')
    print()

Matriz:
[9, 9, 9]
[9, 9, 9]
[9, 9, 9]
[9, 9, 9]
[9, 9, 9]

Matriz:
[9, 9, 9]
[9, 9, 9]
[9, 9, 9]
[9, 9, 9]
[9, 9, 9]

Matriz:
9 9 9 
9 9 9 
9 9 9 
9 9 9 
9 9 9 

Matriz:
9 9 9 
9 9 9 
9 9 9 
9 9 9 
9 9 9 


## Atribuir e ler valores em matrizes:

In [ ]:
# Atribuição direta em posições da matriz

mat = [[0] * 4 for i in range(5)]
mat[0][0] = -1  # linha 0 e coluna 0
mat[2][2] = 99
mat[4][3] = 1000

print('Matriz:')
for lin in mat:
    print(lin)

Matriz:
[-1, 0, 0, 0]
[0, 0, 0, 0]
[0, 0, 99, 0]
[0, 0, 0, 0]
[0, 0, 0, 1000]


In [ ]:
# Leitura de dados em uma matriz

mat = [[0] * 3 for i in range(2)]  # matriz 2x3
for i in range(2):  # for da linha
    for j in range(3):  # for da coluna
        mat[i][j] = int(input(f'Mat[{i}][{j}]: '))
    print()

print('Matriz:')
for lin in mat:
    print(lin)

Mat[0][0]: 5
Mat[0][1]: 4
Mat[0][2]: 6

Mat[1][0]: 4
Mat[1][1]: 5
Mat[1][2]: 4

Matriz:
[5, 4, 6]
[4, 5, 4]


In [ ]:
# Outra forma: usando append

mat = []
for i in range(2):  # for da linha
    lin = []
    for j in range(3):  # for da coluna
        lin.append(int(input(f'Mat[{i}][{j}]: ')))  # leitura da linha
    mat.append(lin)  # insere a linha na matriz
    print()

print('Matriz:')
for lin in mat:
    print(lin)

Mat[0][0]: 5
Mat[0][1]: 4
Mat[0][2]: 6

Mat[1][0]: 4
Mat[1][1]: 7
Mat[1][2]: 8

Matriz:
[5, 4, 6]
[4, 7, 8]


In [ ]:
# Outra forma: usando apenas uma linha

mat = [[int(input(f'Mat[{i}][{j}]: ')) for j in range(3)] for i in range(2)]
print('\nMatriz:')
for lin in mat:
    print(lin)

Mat[0][0]: 4
Mat[0][1]: 5
Mat[0][2]: 4
Mat[1][0]: 8
Mat[1][1]: 8
Mat[1][2]: 7

Matriz:
[4, 5, 4]
[8, 8, 7]


## **Exercícios:**
---
---

### `1.` Ler uma matriz 3x5 de inteiros e, após as leituras, calcular e mostrar:
+ ### A média arimética dos valores da matriz.
+ ### O menor e o maior valor da matriz.
+ ### A qtde de valores negativos contidos na matriz.

In [ ]:
# leitura dos valores da matriz

mat = [[int(input(f'Mat[{i}][{j}]: ')) for j in range(5)] for i in range(3)]

Mat[0][0]: 5
Mat[0][1]: 7
Mat[0][2]: 9
Mat[0][3]: -4
Mat[0][4]: -5
Mat[1][0]: 8
Mat[1][1]: 1
Mat[1][2]: -3
Mat[1][3]: 2
Mat[1][4]: 10
Mat[2][0]: 55
Mat[2][1]: -8
Mat[2][2]: -45
Mat[2][3]: 9
Mat[2][4]: 5


In [ ]:
# cálculo do item 1: somar valor por valor e dividir

soma = 0
for i in range(3):
    for j in range(5):
        soma += mat[i][j]
ma = soma / 15
print(f'Média = {ma:.3f}')

soma = 0
for lin in mat:
    for v in lin:
        soma += v
ma = soma / 15
print(f'Média = {ma:.3f}')

Média = 3.067
Média = 3.067


In [ ]:
# cálculo do item 1: somar linhas e dividir

soma = 0
for lin in mat:
    soma += sum(lin)

ma = soma / 15
print(f'Média = {ma:.3f}')

Média = 3.067


In [ ]:
# usando o módulo (biblioteca) Numpy

import numpy as np

npmat = np.array(mat)

ma = np.average(npmat)  # ou ainda: ma = npmat.sum() / npmat.size
print(f'Média = {ma:.3f}')

Média = 3.067


In [ ]:
# cálculo do item 2

me = ma = mat[0][0]  # menor e maior iniciam com o 1º valor
for lin in mat:
    for v in lin:
        if v < me:
            me = v
        elif v > ma:
            ma = v

print(f'Menor = {me}\n\nMaior = {ma}')

Menor = -45

Maior = 55


In [ ]:
# cálculo do item 2: outro modo

lis_me = [min(lin) for lin in mat]  # cria uma lista com os menores de cada linha
me = min(lis_me)  # calcula o menor da lista de menores
lis_ma = [max(lin) for lin in mat]  # cria uma lista com os maiores de cada linha
ma = max(lis_ma)  # calcula o maior da lista de maiores

print(f'Menor = {me}\n\nMaior = {ma}')

Menor = -45

Maior = 55


In [ ]:
# usando o módulo (biblioteca) Numpy

import numpy as np

npmat = np.array(mat)

me = npmat.min()
ma = npmat.max()

print(f'Menor = {me}\n\nMaior = {ma}')

Menor = -45

Maior = 55


In [ ]:
# Cálculo do item 3

qn = 0
for lin in mat:
    for v in lin:
        if v < 0:
            qn += 1

print(f'Qtde de negativos = {qn}')

Qtde de negativos = 5


In [ ]:
# usando o módulo (biblioteca) Numpy

import numpy as np

npmat = np.array(mat)

# contar através de um filtro
qn = np.count_nonzero(npmat < 0)
print(f'Qtde de negativos = {qn}')

# outra forma: filtrar e contar
qn = npmat[npmat < 0].size
print(f'Qtde de negativos = {qn}')

Qtde de negativos = 5
Qtde de negativos = 5


### `2.` Gerar e mostrar uma matriz 5x5 de inteiros aleatórios entre 0 e 9 e, em seguida, calcular e mostrar a soma dos valores da diagonal principal e o produto dos valores da diagonal secundária da matriz.

In [ ]:
from random import randint

def main():
    mat = [[randint(0, 9) for j in range(5)] for i in range(5)]
    print('Matriz:')
    for lin in mat:
        print(lin)

    s = 0
    for i in range(5):
        s += mat[i][i]
    print(f'\nSoma da diagonal principal = {s}')

    # p = 1
    # j = 4
    # for i in range(5):
    #     p *= mat[i][j]
    #     j -= 1

    p = 1
    for i in range(5):
        p *= mat[i][4 - i]
    print(f'\nProduto da diagonal secundária = {p}')

main()

Matriz:
[8, 9, 7, 0, 3]
[2, 9, 5, 5, 8]
[1, 8, 4, 9, 1]
[3, 4, 7, 6, 7]
[5, 1, 0, 3, 6]

Soma da diagonal principal = 33

Produto da diagonal secundária = 1200


### `3.` Gerar e mostrar uma matriz 3x4 de inteiros aleatórios entre 10 e 30 e, em seguida, calcular e mostrar a qtde de pares e a qtde de ímpares contidos nessa matriz.

In [ ]:
from random import randint
import numpy as np

def main():
    mat = [[randint(10, 30) for j in range(4)] for i in range(3)]
    print('Matriz:')
    for lin in mat:
        print(lin)

    qp = qi = 0
    for lin in mat:
        for v in lin:
            if v % 2 == 0:
                qp += 1
            else:
                qi += 1

    npmat = np.array(mat)
    qp = np.count_nonzero(npmat % 2 == 0)  # ou: npmat[npmat % 2 == 0].size
    qi = npmat.size - qp

    print(f'\nQtde de pares = {qp}')
    print(f'\nQtde de ímpares = {qi}')

main()

Matriz:
[15, 28, 13, 23]
[21, 16, 27, 28]
[11, 18, 10, 15]

Qtde de pares = 5

Qtde de ímpares = 7


### `4.` Gerar e mostrar uma matriz 4x5 com números inteiros e aleatórios entre 0 e 9 e, em seguida, calcular e mostrar:
+ ### O menor e o maior valor de cada linha.
+ ### A média aritmética de cada linha e inseri-la na 6ª coluna. Mostrar a matriz após o cálculo.

In [ ]:
from random import randint

def main():
    mat = [[randint(0, 9) for j in range(5)] for i in range(4)]
    print('Matriz:')
    for lin in mat:
        print(lin)

    for i in range(4):
        print(f'\n{i + 1}ª linha:\nMenor = {min(mat[i])}\nMaior = {max(mat[i])}')

    for lin in mat:
        lin.append(sum(lin) / len(lin))  # ou dividir por 5

    print('\nMatriz com as médias aritméticas de cada linha:')
    for lin in mat:
        print(lin)

main()

Matriz:
[6, 0, 2, 0, 4]
[6, 8, 3, 6, 0]
[5, 4, 5, 4, 4]
[2, 1, 2, 0, 5]

1ª linha:
Menor = 0
Maior = 6

2ª linha:
Menor = 0
Maior = 8

3ª linha:
Menor = 4
Maior = 5

4ª linha:
Menor = 0
Maior = 5

Matriz com as médias aritméticas de cada linha:
[6, 0, 2, 0, 4, 2.4]
[6, 8, 3, 6, 0, 4.6]
[5, 4, 5, 4, 4, 4.4]
[2, 1, 2, 0, 5, 2.0]


### `5.` Gerar uma matriz 5x3 com números inteiros e aleatórios entre 0 e 9 e, em seguida, calcular e mostrar:    
+ ### O menor e o maior valor de cada coluna.
+ ### Calcular e inserir a média aritmética de cada coluna em uma linha no final da matriz. Mostrar a matriz após o cálculo.

In [ ]:
from random import randint

def main():
    mat = [[randint(0, 9) for j in range(3)] for i in range(5)]
    print('Matriz:')
    for lin in mat:
        print(lin)

    medcol = [] # lista com as médias das colunas
    for j in range(3):  # percorre a matriz coluna por coluna
        col = []
        for i in range(5): # linha
            col.append(mat[i][j])
        print(f'\n{j + 1}ª coluna:\nMenor = {min(col)}\nMaior = {max(col)}')
        medcol.append(sum(col) / len(col))  # ou dividir por 3
    mat.append(medcol)

    print('\nMatriz:')
    for lin in mat:
        print(lin)

main()

Matriz:
[5, 1, 5]
[3, 7, 6]
[8, 0, 5]
[8, 8, 9]
[7, 2, 8]

1ª coluna:
Menor = 3
Maior = 8

2ª coluna:
Menor = 0
Maior = 8

3ª coluna:
Menor = 5
Maior = 9

Matriz:
[5, 1, 5]
[3, 7, 6]
[8, 0, 5]
[8, 8, 9]
[7, 2, 8]
[6.2, 3.6, 6.6]


In [ ]:
# usando o módulo (biblioteca) Numpy

from random import randint
import numpy as np

def main():
    mat = [[randint(1, 10) for j in range(3)] for i in range(5)]
    npmat = np.array(mat)

    print('Matriz:')
    print(npmat)

    medcol = []
    for j in range(3):  # coluna
        print(f'\n{j + 1}ª coluna:\nMenor = {npmat[:, j].min()}\nMaior = {npmat[:, j].max()}')
        medcol.append(np.average(npmat[:, j]))

    # inserir uma linha em um array Numpy
    npmat = np.append(npmat, [medcol], axis=0)
    print('\nMatriz com as médias aritméticas de cada coluna:')
    print(npmat)

main()

Matriz:
[[8 5 9]
 [7 7 2]
 [4 8 5]
 [2 9 5]
 [4 7 5]]

1ª coluna:
Menor = 2
Maior = 8

2ª coluna:
Menor = 5
Maior = 9

3ª coluna:
Menor = 2
Maior = 9

Matriz com as médias aritméticas de cada coluna:
[[8.  5.  9. ]
 [7.  7.  2. ]
 [4.  8.  5. ]
 [2.  9.  5. ]
 [4.  7.  5. ]
 [5.  7.2 5.2]]


### `Extra`: inserir uma coluna em um array `Numpy`.

In [ ]:
from random import randint
import numpy as np

def main():
    mat = [[randint(1, 10) for j in range(3)] for i in range(5)]
    npmat = np.array(mat)
    npmat = np.append(npmat, [[99], [99], [99], [99], [99]], axis=1)
    print(npmat)

main()

[[ 1  7  4 99]
 [ 1  1  1 99]
 [ 8  9  2 99]
 [10  2  3 99]
 [10  2  6 99]]


# **Funções (sub-rotinas)**:
---
---
+ ## Blocos de comandos inseridos em um programa e identificados por nomes únicos.
+ ## Executam tarefas específicas dentro de um programa e podem ser chamadas (executadas) sempre que necessário.

## **Tipos de funções:**
---

### **Sem** parâmetros e **sem** retorno:

In [ ]:
def msg():
    print('Ciência da Computação - 2021')

# programa principal

msg()

### **Sem** parâmetros e **com** retorno:

In [ ]:
def soma():
    a = int(input('A: '))
    b = int(input('\nB: '))
    return a + b

# programa principal

s = soma()
print(f'\nSoma = {s}')
if soma() > 10:
    print('Maior que 10')

### **Com** parâmetros e **sem** retorno:

In [ ]:
def potencia(base, exp): # pbase e pexp são parâmetros (dados externos passados à função)
    p = base ** exp
    print(f'{base} elevado a {exp} = {p}')

# programa principal

potencia(2, 10)
potencia(3, 5)
b = 5
e = 8
potencia(b, e)

### **Com** parâmetros e **com** retorno:

In [ ]:
def menor3(a, b, c):
    return min(a, b, c)

# programa principal

def main():
    va = int(input('A: '))
    vb = int(input('\nB: '))
    vc = int(input('\nC: '))
    me = menor3(va, vb, vc)
    print(f'\nMenor = {me}')

    if menor3(va, vb, vc) == va:
        print('\nOK')

main()

## **Exercícios:**
---
---

### `1.` Criar um programa para calcular e mostrar o comprimento e a área de uma circunferência e o volume de uma esfera usando as fórmulas abaixo. Cada cálculo deve ser feito por uma função (3 funções). O main deve ler o raio, chamar as funções e mostrar os resultados no final.
+ ### comp = 2 . PI . raio
+ ### area = PI . raio²
+ ### vol = 4 . PI. raio³ / 3
Onde PI = 3.14

### `2.` Através do exercício anterior, crie apenas uma função para calcular e retornar os 3 cálculos. O restante permanece igual.

### `3.` Criar uma função para calcular e retornar o **Arranjo** e a **Combinação** de `n`, `m` a `m`, onde `n` e `m` são parâmetros da função. No programa principal, `n` e `m` são fornecidos pelo usuário (n >= m), a função é chamada e os resultados são mostrados. Use as fórmulas abaixo para realizar os cálculos:
+ ### A = n! / (n - m)!
+ ### C = n! / (m! . (n - m)!)